src/main/java/io/cdap/plugin/gcp/spanner/sink/SpannerOutputFormat.java

```diff
@@ -231,4 +243,125 @@ public void abortTask(TaskAttemptContext taskAttemptContext) {
+          builder.set(name)
+            .toTimestampArray(
+              timestampList.stream()
+                .map(ts -> Timestamp.ofTimeSecondsAndNanos(ts.toEpochSecond(), ts.getNano()))
+                .collect(Collectors.toList())
+            );
+          break;
+        default:
+          throw new IllegalStateException("Array of '" + logicalType + "' logical type currently not supported.");
+      }
+    } else {
+      Schema.Type type = unionSchema.getType();
+      switch (type) {
+        case BOOLEAN:
+          Collection<Boolean> listBoolean = toCollection(name, type.toString(), values);
+          builder.set(name).toBoolArray(listBoolean);
+          break;
+        case STRING:
+          Collection<String> listString = toCollection(name, type.toString(), values);
+          builder.set(name).toStringArray(listString);
+          break;
+        case LONG:
+          Collection<Long> listValue = toCollection(name, type.toString(), values);
+          builder.set(name).toInt64Array(listValue);
+          break;
+        case DOUBLE:
+          Collection<Double> listDouble = toCollection(name, type.toString(), values);
+          builder.set(name).toFloat64Array(listDouble);
+          break;
+        case INT:
+          Collection<Integer> integerList = toCollection(name, type.toString(), values);
+          builder.set(name)
+            .toInt64Array(
+              integerList.stream()
+                .map(Integer::longValue)
+                .collect(Collectors.toList())
+            );
+          break;
+        case FLOAT:
+          Collection<Float> floatList = toCollection(name, type.toString(), values);
+          builder.set(name)
+            .toFloat64Array(
+              floatList.stream()
+                .map(Float::doubleValue)
+                .collect(Collectors.toList())
+            );
+          break;
+        case BYTES:
+          Collection<byte[]> bytesList = toCollection(name, type.toString(), values);
+          builder.set(name)
+            .toBytesArray(
+              bytesList.stream()
+                .map(ByteArray::copyFrom)
+                .collect(Collectors.toList())
+            );
+          break;
+        default:
+          throw new IllegalStateException("Array of '" + type.name() + "' type currently not supported.");
+      }
+    }
+  }
+
+  private static <T> Collection<T> toCollection(String fieldName, String fieldType, Object value) {
+    Function<String, Collection> valueExtractor = name -> {
+      throw new UnexpectedFormatException(
+        String.format("Field '%s' of type '%s' has unexpected value '%s'", name, fieldType, value));
+    };
+
+    if (value instanceof Collection) {
+      valueExtractor = name -> (Collection<?>) value;
+    } else if (value.getClass().isArray()) {
+      valueExtractor = name -> convertToObjectCollection(value);
+    }
+    return getValue(valueExtractor, fieldName, fieldType, Collection.class);
```

Comment:
this seems really complicated, why not just return the collections in the if-else above?

Reply:
Changed to return Collection, moved to RecordToEntityTransformer class